### ISSUE TITLE
child_process.fork() fails with process exit error code 1 but no other information available

### ISSUE THREAD

**Issue by rfreas** [Azure/iisnode#52]
I am encountering a strange issue with iisnode on Azure.  Basically, the issue is child_process.fork() seems to not work and doesn't throw errors.  I've tried attaching to the uncaughtException event, combing the logs, and so on -- but the only information I get is from the "exit" process event and the error code is 1, which means uncaught exception.  No other information is available.  For the .js file that I am forking, I can run it manually with the exact same command line in the Kudu command line prompt and it runs fine.  But when executed via child_process.fork(), I don't even see the initial console.log() statement execute.  I know the child process is being spawned because I can see it in the Kudu process list with the proper command line and with access to all the same environment variables, and so on.  I am using the latest version of node.exe which is 8.8.4, though initially it was running on whatever the default version was when I setup the original stub template in Azure (node 4.2.4).

Does anyone have thoughts as to what might be going on?

Here's my web.config:

<configuration>

    <system.web>
        <customErrors mode="off" />
    </system.web>
    
    <system.webServer>
        <httpErrors existingResponse="PassThrough" />

        <handlers>
            
            <add name="iisnode" path="index.js" verb="*" modules="iisnode" />
        </handlers>
    
        <rewrite>
            <rules>
                
                <rule name="LogFile" patternSyntax="ECMAScript" stopProcessing="true">
                    <match url="^[a-zA-Z0-9_\-]+\.js\.logs\/\d+\.txt$" />
                </rule>

                
                <rule name="NodeInspector" patternSyntax="ECMAScript" stopProcessing="true">                    
                    <match url="^index.js\/debug[\/]?" />
                </rule>
              
                
                <rule name="StaticContent">
                    <action type="Rewrite" url="public{REQUEST_URI}" />
                </rule>

                
                <rule name="DynamicContent">
                    <conditions>
                        <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True" />
                    </conditions>
                    <action type="Rewrite" url="index.js" />
                </rule>
            </rules>
        </rewrite>
        
        <iisnode
          loggingEnabled="true"
          devErrorsEnabled="true"
         />
    
    </system.webServer>
</configuration>

**Comment by davidebbo** [comment]
Does it repro if you run the same `child_process.fork()` logic from Kudu console? This will tell us whether issue is related to iisnode, or is a more general Azure Web App sandbox issue.

**Comment by rfreas** [comment]
As I explained above, if I manually execute node on the .js file being forked from the Kudu console, the file executes as expected.  When it's being run via child_process.fork(), it doesn't appear to do anything, though I can see the new process spawned from the parent node.exe process in the process list in Kudu.  It's really strange... I tried attaching to process uncaughtException, but I get nothing... the only thing I get is from the exit event on the child process and it's exit code 1, which means uncaughtException.  

Appreciate any insights or steps I might take that would provide additional details as to what is going on.

Thanks,

Rob

**Comment by davidebbo** [comment]
I saw that, but it sounded to me like when you ran it from Kudu, you ran that file directly and did not use fork(), so it's not an equivalent scenario. Please try from Kudu using fork() per my previous reply.

**Comment by rfreas** [comment]
Ahh, I see what you are saying... I misunderstood.  Let me give that a shot and I'll report back -- thanks!

**Comment by rfreas** [comment]
I created a new TestFork.t/js file which just forks the other .js file and I used this command line in the Kudu console:  `"D:\Program Files (x86)\nodejs\8.8.0\node.exe" .\Services\TestFork.js`.  It worked in the sense that the child process was running and I could see logging and other output.  So I wonder what's going on when running in the context of iisnode.  Could there be a permissions issue or something else?  The forked js file downloads files using wget to a temp folder (d:\home\site\temp, the app creates it if it doesn't exist during start up), so there's some read/write file i/o but that seems to run fine as I can see the files being created and later deleted via fs.unlinkSync(), etc.

If there's really an uncaughtException as error code 1 suggests, how come I can't seem to catch it in a try/catch around the child_process.fork() call, or by listening to process.on uncaughtException or child.on uncaughtException, etc.?

What do you think?  Thanks again for your assistance.

**Comment by davidebbo** [comment]
> The forked js file downloads files using wget to a temp folder (d:\home\site\temp, the app creates it if it doesn't exist during start up), so there's some read/write file i/o but that seems to run fine as I can see the files being created and later deleted via fs.unlinkSync(), etc.

You mean you see that happening in both cases or just in the Kudu console case?

I'd say the next step is to isolate whether the issue is more:

1. Something related to fork() not working in the context of iisnode
1. fork() working fine but something in the specific logic that you run in the forked file not working

You should be able to isolate that by trying to fork() a trivial .js file instead of the more complex one.

**Comment by rfreas** [comment]
I see it happening in the Kudu console case for sure, but it's not clear when the same file is forked.  I don't see anything from stdout -- console.log() statements don't show up in the forked case.  I added the `{ silent: true}` option, which is supposed to pipe output back to the parent process, but I still don't see any output.

I tried a fork() on a trivial .js that had nothing in it but a console.log() statement and the process exit code was 0--which is good.  However, again, I did not see any output and I am not sure why.

I think my issue is proposition 2 that you put forth above, but without being able to see any output or any feedback whatsoever from the child process, it is more difficult to diagnose.  Any idea why I am not seeing console.log() statements when the only option I pass to fork() is `silent:true`?

I also had a thought that the command line parameters on Windows weren't passing properly to the forked process when I saw there is a `windowsVerbatimArguments` option however, this option must be a new one as it doesn't exist in the node version we're using on the client--which is 7.0.39.  I may try upgrading both to the latest and see if the problems persist.  I've bounced around versioning trying different things, but presently I am using 7.10.0 on the server which is the closest available approximation to 7.0.39 on the client.

**Comment by davidebbo** [comment]
@rfreas can you share both test files that you are using from Kudu?

- the one that calls fork()
- the simplified one that fork starts

**Comment by rfreas** [comment]
Here are the two files.  This works fine from KUDU, however, I still don't see any output from the console.log() statements in MultiDocumentsDownloadWorker.  Process exit code is 0.

[TestFork.txt](https://github.com/Azure/iisnode/files/1856866/TestFork.txt)
[MultiDocumentsDownloadWorker.txt](https://github.com/Azure/iisnode/files/1856867/MultiDocumentsDownloadWorker.txt)

Does child_process.fork() handle the interception process of iisnode?  i.e., in the KUDU process explorer, I see this is the command line for the main entry point: `"D:\Program Files (x86)\nodejs\7.10.1\node.exe" "D:\Program Files (x86)\iisnode\interceptor.js" "D:\home\site\wwwroot\index.js"`  ... how does that work with a fork()?  There's no interception via iisnode.

**Comment by davidebbo** [comment]
So if I read you correctly, you are saying that even from Kudu Console (i.e no iisnode in the picture), you don't get any console output. If so, can we just focus on this part and leave iisnode out if the equation?

**Comment by davidebbo** [comment]
I just tried your test code on my local machine, and I don't see the Console log coming from MultiDocumentsDownloadWorker.js there either. So this appears to be normal Node behavior and not related to Kudu or iisnode.

**Comment by rfreas** [comment]
Gotcha.  Hmm... so how do I get more information to know where the problem is in the child js file when forked?  No exception is thrown nor is the uncaughtException event fired.

**Comment by davidebbo** [comment]
I guess you could try logging to a file instead. There is probably a simpler way, but I'm no Node expert.

**Comment by rfreas** [comment]
So, as it turns out, what I needed to do was use process.send() to relay debugging information I previously was logging with console.log(), and then from the parent process, attach to the child process' 'message' event in order to gain insight into the problems the script was encountering, which were several fold.

I also made sure my node versions matched (both iisnode and my code targeting v8.9.4) which corrected an issue around command line arguments being quoted improperly.

The real issue was not having insight into what was going on in the child process--using process.send() helped shed light on this.

Thanks again @davidebbo.

Rob
